cmd/query/app/querysvc/adjuster/clockskew.go

```diff
@@ -0,0 +1,203 @@
+// Copyright (c) 2024 The Jaeger Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+package adjuster
+
+import (
+	"fmt"
+	"time"
+
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+
+	"github.com/jaegertracing/jaeger/internal/jptrace"
+	"github.com/jaegertracing/jaeger/pkg/otelsemconv"
+)
+
+const (
+	warningDuplicateSpanID     = "duplicate span IDs; skipping clock skew adjustment"
+	warningMissingParentSpanID = "parent span ID=%s is not in the trace; skipping clock skew adjustment"
+	warningMaxDeltaExceeded    = "max clock skew adjustment delta of %v exceeded; not applying calculated delta of %v"
+	warningSkewAdjustDisabled  = "clock skew adjustment disabled; not applying calculated delta of %v"
+)
+
+// ClockSkew returns an Adjuster that corrects span timestamps for clock skew.
+//
+// This adjuster modifies the start and log timestamps of child spans that are
+// inconsistent with their parent spans due to clock differences between hosts.
+// It assumes all spans have unique IDs and should be used after SpanIDUniquifier.
+//
+// The adjuster determines if two spans belong to the same source by deriving a
+// unique string representation of a host based on resource attributes.
+// Specifically, it checks the "ip" attribute of the resource.
+// If two spans have the same host key, they are considered to be from
+// the same source, and no clock skew adjustment is expected between them.
+//
+// Parameters:
+//   - maxDelta: The maximum allowable time adjustment. Adjustments exceeding
+//     this value will be ignored.
+func ClockSkew(maxDelta time.Duration) Adjuster {
+	return Func(func(traces ptrace.Traces) {
+		adjuster := &clockSkewAdjuster{
+			traces:   traces,
+			maxDelta: maxDelta,
+		}
+		adjuster.buildNodesMap()
+		adjuster.buildSubGraphs()
+		for _, root := range adjuster.roots {
+			skew := clockSkew{hostKey: root.hostKey}
+			adjuster.adjustNode(root, nil, skew)
+		}
+	})
+}
+
+type clockSkewAdjuster struct {
+	traces   ptrace.Traces
+	maxDelta time.Duration
+	spans    map[pcommon.SpanID]*node
+	roots    map[pcommon.SpanID]*node
+}
+
+type clockSkew struct {
+	delta   time.Duration
+	hostKey string
+}
+
+type node struct {
+	span     ptrace.Span
+	children []*node
+	hostKey  string
+}
+
+// hostKey derives a unique string representation of a host based on resource attributes.
+// This is used to determine if two spans are from the same host.
+func hostKey(resource ptrace.ResourceSpans) string {
+	if attr, ok := resource.Resource().Attributes().Get(string(otelsemconv.HostIDKey)); ok {
+		return attr.Str()
+	}
+	if attr, ok := resource.Resource().Attributes().Get(string(otelsemconv.HostIPKey)); ok {
+		ips := attr.Slice()
```

Comment:
I suspect some people may ignore setting a slice and set just a plain string, so I would handle that use case too.

Reply:
@yurishkuro is it okay to break the contract of the semantic convention here?